storage/aptosdb/src/fast_sync_aptos_db.rs

```diff
@@ -0,0 +1,1110 @@
+// Copyright © Aptos Foundation
+// Parts of the project are originally copyright © Meta Platforms, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    backup::backup_handler::DbState,
+    event_store::EventStore,
+    ledger_store::LedgerStore,
+    metrics::{
+        BACKUP_EPOCH_ENDING_EPOCH, BACKUP_STATE_SNAPSHOT_LEAF_IDX, BACKUP_STATE_SNAPSHOT_VERSION,
+        BACKUP_TXN_VERSION,
+    },
+    state_store::StateStore,
+    transaction_store::TransactionStore,
+    AptosDB,
+};
+use anyhow::{anyhow, ensure, format_err, Context, Result};
+use aptos_config::config::{BootstrappingMode, NodeConfig};
+use aptos_crypto::HashValue;
+use aptos_storage_interface::{
+    cached_state_view::ShardedStateCache, state_delta::StateDelta, DbReader, DbWriter,
+    ExecutedTrees, FastSyncStatus, Order, StateSnapshotReceiver,
+};
+use aptos_types::{
+    account_config::NewBlockEvent,
+    contract_event::{ContractEvent, EventWithVersion},
+    epoch_change::EpochChangeProof,
+    epoch_state::EpochState,
+    event::EventKey,
+    ledger_info::LedgerInfoWithSignatures,
+    proof::{
+        AccumulatorConsistencyProof, SparseMerkleProof, SparseMerkleProofExt,
+        SparseMerkleRangeProof, TransactionAccumulatorRangeProof, TransactionAccumulatorSummary,
+        TransactionInfoWithProof,
+    },
+    state_proof::StateProof,
+    state_store::{
+        state_key::StateKey,
+        state_key_prefix::StateKeyPrefix,
+        state_storage_usage::StateStorageUsage,
+        state_value::{StateValue, StateValueChunkWithProof},
+        table::{TableHandle, TableInfo},
+        ShardedStateUpdates,
+    },
+    transaction::{
+        AccountTransactionsWithProof, Transaction, TransactionInfo, TransactionListWithProof,
+        TransactionOutputListWithProof, TransactionToCommit, TransactionWithProof, Version,
+    },
+    write_set::WriteSet,
+};
+use move_core_types::account_address::AccountAddress;
+use std::sync::{Arc, RwLock};
+
+const SECONDARY_DB_DIR: &str = "fast_sync_secondary";
+
+/// This is a wrapper around [AptosDB] that is used to bootstrap the node for fast sync mode
+pub struct FastSyncStorageWrapper {
+    // main is used for normal read/write or genesis data during fast sync
+    db_main: AptosDB,
```

Comment:
Interesting... some thoughts:

1. Nit: I'd likely call these something more obvious, e.g.: `temporary_db_with_genesis` and `real_db_for_fast_sync`?
2. I probably lack context, but do these types need to be `AptosDB`? Why can they not be `DBReaderWriter`? Is this because of the backup system?
3. If we only use this wrapper in the context of fast sync, you can avoid the `Option<AptosDB>` (for the second DB) and clean up the unwraps below?
4. I don't really know why we need `FastSyncStatus`? Could we not just have a `fast_sync_complete: AtomicBool`? This would avoid the need for grabbing a RWLock for all reads and writes, and simplify the switches (i.e., the `FastSyncStatus::UNKNOWN` and `FastSyncStatus::STARTED` doesn't mean much, right? All we need to know is when it finishes, so we can direct all calls to `real_db_for_fast_sync`?) Does that make sense?

Reply:
1. that was my previous implementation. But there is a corner case. For example, the node can be configured to do fast sync. When all its peers only have genesis data, it would actually do a continuous update of txns one by one. Storage in this case should put the genesis and following updates in the same DB.  
2. Yes, this is related to the backup system since it needs to access the underlying txn_store, event_store, etc of the DB. 
3. I don't know if that makes the logic more complicated. since you need to remember AptosDB is used in this circumstance and FastSyncDBWrapper is used in other circumstances. For now, I just make FastSyncDBWrapper used in all circumstances and people can check one place to find the code behavior.
4. as mentioned in 1. we need to know if fast sync indeed starts to determine if we should use a separate DB. the `Unknown` is used for indicating non-fast-sync cases.

Reply:
1. I'd need to look at the state sync code again, but if that's the case, I think we should change state sync such that it will always fast sync (i.e., if it starts up in fast sync mode, and we haven't had a single epoch change, state sync will just loop until the first epoch change occurs). Otherwise, this feels a bit janky to me.
3 + 4. Not really following you -- will sync offline 😄

Reply:
updated based on offline discussion. The test will be correct once the state sync change is landed.